File: 2024/day02/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func readInput(filename string) ([][]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}

	var reports [][]int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parts := strings.Fields(scanner.Text())
		values := make([]int, len(parts))
		for i, p := range parts {
			v, err := strconv.ParseInt(p, 10, 64)
			if err != nil {
				return nil, err
			}
			values[i] = int(v)
		}
		reports = append(reports, values)
	}
	return reports, nil
}

func isSafe(report []int) bool {
	deltas := make([]int, len(report)-1)
	for i := 1; i < len(report); i++ {
		deltas[i-1] = report[i] - report[i-1]
	}
	c := 1
	if deltas[0] < 0 {
		c = -1
	}
	for _, d := range deltas {
		d2 := c * d
		if d2 < 1 || d2 > 3 {
			return false
		}
	}
	return true
}

func solve(reports [][]int) (int, int) {
	safeCt := 0
	safeCtWithDampener := 0
	for _, report := range reports {
		if isSafe(report) {
			safeCt += 1
		} else {
			for i := 0; i < len(report); i++ {
				dampenedReport := make([]int, 0)
				dampenedReport = append(dampenedReport, report[:i]...)
				dampenedReport = append(dampenedReport, report[i+1:]...)
				if isSafe(dampenedReport) {
					safeCtWithDampener += 1
					break
				}
			}
		}
	}
	return safeCt, safeCt + safeCtWithDampener
}

func main() {
	filename := "input.txt"
	if len(os.Args) == 2 {
		filename = os.Args[1]
	}
	if len(os.Args) > 2 {
		log.Fatalf("%s: <filename>", os.Args[0])
	}

	reports, err := readInput(filename)
	if err != nil {
		log.Fatal("Error:", err)
	}

	safeCt, safeCtWithDampener := solve(reports)
	fmt.Printf("Part1: safe_ct = %d\n", safeCt)
	fmt.Printf("Part2: safe_ct_with_dampener = %d\n", safeCtWithDampener)
}

File: 2024/day02/main.rs
#![feature(array_windows)]
use std::{env, fs};

fn is_safe(report: &[i32]) -> bool {
    let mut deltas = report.array_windows::<2>()
        .map(|pair| pair[1] - pair[0])
        .collect::<Vec<_>>();
    if deltas[0] < 0 {
        deltas = deltas.into_iter().map(|v| -1 * v).collect();
    }

    for &d in &deltas {
        if d < 1 || d > 3 {
           return false;
        }
    }
    true
}

fn solve(reports: &Vec<Vec<i32>>) -> (i32, i32) {
    let mut safe_ct = 0;
    let mut safe_ct_with_dampening = 0;
    for report in reports {
        if is_safe(&report) {
            safe_ct += 1;
        } else {
            for i in 0..report.len() {
                let mut dampened_report = report.clone();
                dampened_report.remove(i);
                if is_safe(&dampened_report) {
                    safe_ct_with_dampening += 1;
                    break;
                }
            }
        }
    }
    (safe_ct, safe_ct + safe_ct_with_dampening)
}
fn read_input(filename: &str) -> Vec<Vec<i32>> {
    fs::read_to_string(filename).expect("Unable to open file")
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|v| v.parse::<i32>().unwrap())
                .collect::<Vec<_>>()
        }).collect()
}

fn main() {
    let mut filename = "input.txt";
    let args: Vec<String> = env::args().collect();
    if args.len() == 2 {
        filename = &args[1];
    }

    let reports = read_input(filename);
    let (safe_ct, safe_ct_with_dampening) = solve(&reports);
    println!("Part1: safe_ct = {}", safe_ct);
    println!("Part2: safe_ct_with_dampening = {}", safe_ct_with_dampening);
}